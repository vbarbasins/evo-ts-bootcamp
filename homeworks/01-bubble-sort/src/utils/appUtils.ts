type SortingDoneResult = {
  newNumbers: number[],
  nextSortingStep: {
    iteration: number,
    currentIndex: number,
  }
}

type SortingStepResult = SortingDoneResult | null

export const makeSortingStep = (
  numbers: number[],
  iteration: number,
  currentIndex: number,
): SortingStepResult => {
  if (iteration < numbers.length) {
    const newNumbers = [...numbers];
    const nextIndex = currentIndex + 1;
    if (numbers[currentIndex] > numbers[nextIndex]) {
      const tmp = newNumbers[currentIndex];
      newNumbers[currentIndex] = newNumbers[nextIndex];
      newNumbers[nextIndex] = tmp;
    }
    const continueIteration = nextIndex < numbers.length - iteration;
    return {
      newNumbers,
      nextSortingStep: {
        iteration: continueIteration ? iteration : iteration + 1,
        currentIndex: continueIteration ? nextIndex : 0,
      },
    };
  }
  return null;
};

export const createRandomNumberArray = (
  maxNumber: number,
  numberCount: number,
): number[] => Array.from(
  { length: numberCount },
  () => Math.floor(Math.random() * maxNumber) + 1,
);
