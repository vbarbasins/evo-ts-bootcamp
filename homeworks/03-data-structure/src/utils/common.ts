export const assertNever = (arg: never): never => {
  throw new Error(`Unexpected argument: ${arg}`);
};
