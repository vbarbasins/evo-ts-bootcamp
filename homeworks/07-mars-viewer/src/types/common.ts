export type Photo = {
  id: number;
  // eslint-disable-next-line camelcase
  img_src: string;
};

export type SolPhotoSet = { photoSet: Photo[], sol: number };

export type AppState = {
  favourites: Photo[];
  currentSol: number;
  solPhotoSets: SolPhotoSet[];
}
