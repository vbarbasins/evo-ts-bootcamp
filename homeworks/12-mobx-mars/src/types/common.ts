export type NasaPhoto = {
  id: number;
  // eslint-disable-next-line camelcase
  img_src: string;
}

export type Photo = NasaPhoto & {
  favourite: boolean;
  sol: number;
};

export type AppState = {
  currentSol: number;
  photos: Photo[];
  loadingPhotos: boolean;
  showingFavourites: boolean;
}
