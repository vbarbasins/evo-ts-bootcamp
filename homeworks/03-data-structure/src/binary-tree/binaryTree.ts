import { assertNever } from '../utils/commonUtils';

export type TreeNode<T> = {
  value: T;
  left: TreeNode<T> | null;
  right: TreeNode<T> | null;
};

export enum TraverseType {
  depthFSInOrder = 'DFS_INORDER',
  depthFSPreOrder = 'DFS_PREORDER',
  depthFSPostOrder = 'DFS_POSTORDER',
  breadthFS = 'BFS',
}

export interface Tree<T> {
  setTreeNode(treeNode: TreeNode<T>): this;
  traverse(type: TraverseType): T[];
  getColumn(column: number): T[];
}

export class BinaryTree<T> implements Tree<T> {
  protected rootNode: TreeNode<T>;

  constructor(treeNode: TreeNode<T>) {
    this.rootNode = treeNode;
  }

  public setTreeNode(treeNode: TreeNode<T>): this {
    this.rootNode = treeNode;
    return this;
  }

  public getColumn(targetColumn: number): T[] {
    const { value, left, right } = this.rootNode;
    const currentColumn = 0;
    const result: T[] = [];
    if (currentColumn === targetColumn) result.push(value);
    const traverseLeft = left === null ? [] : new BinaryTree(left).getColumn(targetColumn + 1);
    const traverseRight = right === null ? [] : new BinaryTree(right).getColumn(targetColumn - 1);
    return [...traverseLeft, ...result, ...traverseRight];
  }

  public traverse(type: TraverseType): T[] {
    const { value, left, right } = this.rootNode;
    const traverseLeft = left === null ? [] : new BinaryTree(left).traverse(type);
    const traverseRight = right === null ? [] : new BinaryTree(right).traverse(type);

    switch (type) {
      case TraverseType.breadthFS: {
        const result: T[] = [];
        const queue: TreeNode<T>[] = [this.rootNode];
        while (queue.length > 0) {
          const node = queue.shift()!;
          result.push(node.value);
          if (node.left) queue.push(node.left);
          if (node.right) queue.push(node.right);
        }
        return result;
      }
      case TraverseType.depthFSInOrder: return [...traverseLeft, value, ...traverseRight];
      case TraverseType.depthFSPreOrder: return [value, ...traverseLeft, ...traverseRight];
      case TraverseType.depthFSPostOrder: return [...traverseLeft, ...traverseRight, value];
      default: return assertNever(type);
    }
  }
}
