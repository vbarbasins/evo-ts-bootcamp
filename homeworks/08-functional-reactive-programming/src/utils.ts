import { GRID_CELL_COUNT, INITIAL_WINDOWS } from './consts';
import { Cell, CellContent } from './types';

export const generateGrid = (): Cell[] => Array.from(
  { length: GRID_CELL_COUNT },
  (v, n) => ({
    id: n,
    content: INITIAL_WINDOWS.includes(n) ? CellContent.WINDOW : CellContent.WALL,
  }),
);

export const getRandomTarget = (): number => (
  INITIAL_WINDOWS[Math.floor(Math.random() * INITIAL_WINDOWS.length)]
);
