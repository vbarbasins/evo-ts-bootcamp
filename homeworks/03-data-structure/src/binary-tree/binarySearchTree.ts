import { Tree, BinaryTree, TraverseType } from './binaryTree';

export interface SearchTree extends Tree<number> {
  has(value: number): boolean;
}

export class BinarySearchTree extends BinaryTree<number> implements SearchTree {
  has(value: number): boolean {
    return this.traverse(TraverseType.BFS).includes(value);
  }
}
