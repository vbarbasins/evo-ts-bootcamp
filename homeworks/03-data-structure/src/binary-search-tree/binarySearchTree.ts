import { Tree, BinaryTree, TraverseType } from '../binary-tree/binaryTree';

export interface SearchTree extends Tree<number> {
  has(value: number): boolean;
}

export class BinarySearchTree extends BinaryTree<number> implements SearchTree {
  public has(value: number): boolean {
    return this.traverse(TraverseType.breadthFS).includes(value);
  }
}
