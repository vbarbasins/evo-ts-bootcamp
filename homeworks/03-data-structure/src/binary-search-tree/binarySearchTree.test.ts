import { TraverseType } from '../binary-tree/binaryTree';
import { assertNever } from '../utils/common';
import { BinarySearchTree } from './binarySearchTree';

jest.mock('../utils/common');

/*             43
*             /  \
*          40     46
*         /  \   /  \
*       10    2 1    5
*              \
*               67
*/
const givenSearchTree = new BinarySearchTree({
  value: 43,
  left: {
    value: 40,
    left: { value: 10, left: null, right: null },
    right: {
      value: 2,
      left: null,
      right: { value: 67, left: null, right: null },
    },
  },
  right: {
    value: 46,
    left: { value: 1, left: null, right: null },
    right: { value: 5, left: null, right: null },
  },
});

describe('BinarySearchTree', () => {
  describe('has', () => {
    describe.each([
      [67, givenSearchTree, true],
      [44, givenSearchTree, false],
      [1, givenSearchTree, true],
    ])('.has(%i)', (value, given, expected) => {
      it(`should return ${expected}`, () => {
        expect(given.has(value)).toEqual(expected);
      });
    });
  });

  describe('setTreeNode', () => {
    it('should set new tree node as root node', () => {
      const tree = new BinarySearchTree({ value: 1, left: null, right: null });
      tree.setTreeNode({ value: 2, left: null, right: null });
      expect(tree).toMatchSnapshot();
    });
  });

  describe('traverse', () => {
    it('should call assertNever from utils, when passed type is not handled', () => {
      const given = new BinarySearchTree({ value: 1, left: null, right: null });
      const mockAssertNever = assertNever as jest.MockedFunction<typeof assertNever>;
      given.traverse('some type' as TraverseType);
      expect(mockAssertNever).toHaveBeenCalled();
    });

    describe.each([
      [TraverseType.depthFSInOrder, givenSearchTree, [10, 40, 2, 67, 43, 1, 46, 5]],
      [TraverseType.depthFSPreOrder, givenSearchTree, [43, 40, 10, 2, 67, 46, 1, 5]],
      [TraverseType.depthFSPostOrder, givenSearchTree, [10, 67, 2, 40, 1, 5, 46, 43]],
      [TraverseType.breadthFS, givenSearchTree, [43, 40, 46, 10, 2, 1, 5, 67]],
    ])('.traverse(%p)', (traverseType, given, expected) => {
      it(`should return ${expected}`, () => {
        expect(given.traverse(traverseType)).toStrictEqual(expected);
      });
    });
  });

  describe('getColumn', () => {
    describe.each([
      [0, givenSearchTree, [2, 43, 1]],
      [-1, givenSearchTree, [40]],
      [1, givenSearchTree, [67, 46]],
      [-3, givenSearchTree, []],
    ])('.getColumn(%i)', (column, given, expected) => {
      it(`should return ${expected}`, () => {
        expect(given.getColumn(column)).toStrictEqual(expected);
      });
    });
  });
});
