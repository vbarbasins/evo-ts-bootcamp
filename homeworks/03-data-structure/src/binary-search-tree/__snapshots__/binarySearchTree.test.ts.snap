// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinarySearchTree setTreeNode should set new tree node as root node 1`] = `
BinarySearchTree {
  "rootNode": Object {
    "left": null,
    "right": null,
    "value": 2,
  },
}
`;
